packages/cursorless-engine/src/languages/TreeSitterQuery/queryPredicateOperators.ts

```diff
@@ -80,10 +81,66 @@ class End extends QueryPredicateOperator<End> {
   }
 }
 
+class ChildRange extends QueryPredicateOperator<ChildRange> {
+  name = "child-range!" as const;
+  schema = z.union([
+    z.tuple([q.node, q.integer]),
+    z.tuple([q.node, q.integer, q.integer]),
+    z.tuple([q.node, q.integer, q.integer, q.boolean]),
+    z.tuple([q.node, q.integer, q.integer, q.boolean, q.boolean]),
+  ]);
+
+  run(nodeInfo: MutableQueryCapture, startIndex: number): boolean;
+  run(
+    nodeInfo: MutableQueryCapture,
+    startIndex: number,
+    endIndex: number,
+  ): boolean;
+  run(
+    nodeInfo: MutableQueryCapture,
+    startIndex: number,
+    endIndex: number,
+    excludeStart: boolean,
+  ): boolean;
+  run(
+    nodeInfo: MutableQueryCapture,
+    startIndex: number,
+    endIndex: number,
+    excludeStart: boolean,
+    excludeEnd: boolean,
+  ): boolean;
+  run(
+    nodeInfo: MutableQueryCapture,
+    startIndex: number,
+    endIndex?: number,
+    excludeStart?: boolean,
+    excludeEnd?: boolean,
+  ) {
```

Comment:
If I understand correctly the arguments are defined by the schema. In which case this is the only way we could do that.

Alternatives would be an single object arg, but the types are probably just as hard.

The easiest but least typesafe way might be to make the schema a bit looser to match the last function args(ie change the schema to use optional or something, I'm not familiar with Zoe yet), but I'm guessing we want the types to be strict.

Reply:
Yeah the problem with making them optional (zod does support that) is that then it complains when the list it's validating doesn't have enough args.  For more context, we're using this schema to validate query predicate operators in tree-sitter queries.  See eg https://github.com/cursorless-dev/cursorless/pull/1508/files#diff-648955c590f6bd110d61e822e11d4e3247d92dfc3a9843426680b190bc7ff0b5R9

I also don't think object will work, because they just come in as an ordered list with no names https://github.com/cursorless-dev/cursorless/blob/pokey/support-jsx-tag-scopes-using-tree-sitter-queries/packages/cursorless-engine/src/languages/TreeSitterQuery/QueryPredicateOperator.ts#L60

Reply:
It looks like there might be a convoluted zod workaround with the second argument of tuple. It seems that there's a few questions and issues about this particular problem on their repo. https://github.com/colinhacks/zod/issues/149

I think best bet is to try get zod to accept the optional arguments - otherwise this is probably the best we can do.

Reply:
Ok I made it work by just removing the type on `queryPredicateOperators` at the bottom of the file 🤷‍♂️.  That leaves things just as strongly typed, so I'm happy